Fix Py_None reference counting in pypp, prevent modification of RunSingleTest/CMakeLists.txt, check for TODO comments

nilsdeppe opened the pull request:
## Proposed changes

This caused the fun and easy to understand error:

Fatal Python error: deallocating None

The backtrace was basically useless ("something went wrong with the python call,
good luck!")

There's not really a good way to test for this. I only discovered this by making a dozen or more calls into pypp with `std::optional`. 

Fix #549

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Fix Py_None reference counting in pypp

This caused the fun and easy to understand error:

Fatal Python error: deallocating None

The backtrace was basically useless ("something went wrong with the python call,
good luck!")

We use Py_RETURN_NONE since that does the correct reference counting for us, and
also allows us to check for 'return Py_None;' in CI.

kidder commented on tests/Unit/Framework/PyppFundamentals.hpp:
```diff
@@ -87,7 +87,10 @@ PyObject* make_py_tuple(const Args&... t) {
 struct ToPyObject<void, std::nullptr_t> {
-  static PyObject* convert() { return Py_None; }
+  static PyObject* convert() {
+    Py_INCREF(Py_None);
```
you can if you wish combine these two lines by using `Py_RETURN_NONE;`

nilsdeppe replied:
I think this is clearer, actually. I don't love a function that returns something having only:
```cpp
PyObject* convert() {
  Py_RETURN_NONE;
}
```
I find the lack of an actual return statement makes it harder to read :shrug: But I don't dislike it enough to fight over it. Whatever you prefer :)

kidder replied:
I don't care either way as, starting in the not too distant future, I'd likely have to look up either one to understand what is going on...  

An argument for preferring `Py_RETURN_NONE` is that we could add a CI file test for not having `return Py_None;` to prevent a future bug...

nilsvu replied:
I like the idea of preventing this bug with a file check. FYI, this is the definition of `Py_RETURN_NONE`:

```c++
#define Py_RETURN_NONE return Py_NewRef(Py_None)
```
https://github.com/python/cpython/blob/8b6c4a921af6d5d0a9640211ac93d7886a55a8f3/Include/object.h#L563

nilsdeppe replied:
Yea, add a CI check is a good argument for using `Py_RETURN_NONE`. I'll adjust

kidder reviewed (changes requested):
Oh, could you add a bullet to the list for `CHECK_FILES` in `docs/DevGuide/Travis.md` listing the check and the suggested fix

nilsdeppe committed:
Prevent changing RunSingleTest/CMakeLists.txt

nilsdeppe commented:
Sure, I'ma lso adding a check that `RunSingleTests/CMakeLists.txt` isn't getting modified

fmahebert commented on docs/DevGuide/Travis.md:
```diff
@@ -69,13 +69,16 @@ if any of the following are true:
     * contains `Ls` (use `List` instead)
     * contains additional text after `/*!` (does not render correctly in
       Doxygen)
+    * contains the string `return Py_None;` (bug prone, use `Py_RETUR_NONE`
```
RETUR -> RETURN

fmahebert commented on docs/DevGuide/Travis.md:
```diff
@@ -69,13 +69,16 @@ if any of the following are true:
     * contains `Ls` (use `List` instead)
     * contains additional text after `/*!` (does not render correctly in
       Doxygen)
+    * contains the string `return Py_None;` (bug prone, use `Py_RETUR_NONE`
```
RETUR -> RETURN

nilsdeppe replied:
:clinking_glasses:

nilsvu commented on tools/FileTestDefs.sh:
```diff
@@ -418,6 +418,35 @@ pragma_once_test() {
+        staged_grep -q -x '.*return Py_None;.*' "$1"
+}
+py_return_none_report() {
+    echo "Did not find 'return Py_None;' in these files:"
```
This should probably say the opposite :) and could suggest to use `Py_RETURN_NONE` instead

nilsdeppe committed:
Prevent TODO comments in CI

We should be filing issues instead of adding TODO comments. However, using TODO
comments in local development is reasonable.

nilsdeppe commented:
I also added a commit that prevents TODO comments from being added to the code. We should be filing issues instead. What this does not check for are `\todo` Doxygen comments. Do we want to prevent those as well? Currently we have set Doxygen not to generate a `TODO` list

kidder commented:
@nilsleiffischer please look at

nilsvu commented on tools/FileTestDefs.sh:
```diff
@@ -547,10 +547,32 @@ prevent_cpp_includes_test() {
-    run_tests "${standard_checks[@]}"
+if [ "$1" = --test ] ; then
+    checks_for_test=${standard_checks}
+    checks_for_test+=(prevent_run_single_test_changes)
```
Could you add a comment here please, to describe why this special case is necessary?

nilsvu commented on tools/Hooks/pre-commit.sh:
```diff
@@ -28,7 +28,9 @@ found_error=0
 
-printf '%s\0' "${commit_files[@]}" | run_checks "${standard_checks[@]}"
+commit_hook_checks=${standard_checks}
+commit_hook_checks+=(prevent_run_single_test_changes)
```
also a comment here would be nice

wthrowe commented on tools/FileTestDefs.sh:
```diff
@@ -518,10 +549,32 @@ prevent_cpp_includes_test() {
-    run_tests "${standard_checks[@]}"
+if [ "$1" = --test ] ; then
+    checks_for_test=${standard_checks}
+    checks_for_test+=(prevent_run_single_test_changes)
```
I don't think this works.  (From experiment, I think you lose all but the first standard check.)  A simpler way to do this is
```bash
[ "$1" = --test ] && staged_grep() { grep "$@"; } && \
￼    run_tests "${standard_checks[@]}" prevent_run_single_test_changes
```

nilsdeppe replied:
Ah thanks! I tried that initially with `"${standard_checks[@]} prevent..."`. I was going to ask you about this :)

nilsdeppe commented:
I've rebased and squashed in the changes since they were about the same number of lines as the commits were to start with.

I added the explanation as to why the run single test check isn't in `standard_checks` to the comment above the test itself. I think that's the most appropriate place for it, personally.

kidder reviewed (approved):
@wthrowe please look at

wthrowe reviewed (approved):
> `"${standard_checks[@]} prevent..."`

Yeah, putting an array expansion in quotes along with something else does something bizarre and useless.  (I forget the details.)
